src/gt4py/next/program_processors/runners/dace_iterator/__init__.py

```diff
@@ -63,14 +63,29 @@ def convert_arg(arg: Any):
     return arg
 
 
-def preprocess_program(program: itir.FencilDefinition, offset_provider: Mapping[str, Any]):
-    program = apply_common_transforms(
+def preprocess_program(
+    program: itir.FencilDefinition, offset_provider: Mapping[str, Any], lift_mode: LiftMode
+):
+    node = apply_common_transforms(
         program,
-        offset_provider=offset_provider,
-        lift_mode=LiftMode.FORCE_INLINE,
         common_subexpression_elimination=False,
+        lift_mode=lift_mode,
+        offset_provider=offset_provider,
+        unroll_reduce=False,
     )
-    return program
+    # If we don't unroll, there may be lifts left in the itir which can't be lowered to SDFG.
+    # In this case, just retry with unrolled reductions.
+    if all([ItirToSDFG._check_no_lifts(closure) for closure in node.closures]):
+        fencil_definition = node
+    else:
+        fencil_definition = apply_common_transforms(
+            program,
+            common_subexpression_elimination=False,
+            lift_mode=lift_mode,
+            offset_provider=offset_provider,
+            unroll_reduce=True,
+        )
```

Comment:
Can you explain why you don't unroll in the first place?

Reply:
This is the approach used in GTFN backend. We could play with this option to see what SDFG representation is most suitable for DaCe optimizations. The non-unrolled version produces a more compact SDFG.

Reply:
I think my question was more like, can you always generate the DaCe SDFG with `unroll_reduce=False`? Because you're first running with `False`, then check for lifts, then run with `True`, which does not look too robust.

As for optimizations, I agree with you, we can play around later.